File: stacks/node/base/code/task-trace.tpl.js
const fs = require('fs');
const path = require('path');

function tracePath() {
  return process.env.TASK_TRACE_PATH || path.join('artifacts', 'task-trace.jsonl');
}

function writeEvent(ev) {
  const p = tracePath();
  fs.mkdirSync(path.dirname(p), { recursive: true });
  const event = { ...ev, ts: ev.ts ?? Date.now() / 1000 };
  fs.appendFileSync(p, JSON.stringify(event) + '\n', { encoding: 'utf8' });
}

function emit({ task_id, type, name, key, table, keys, value, fields }) {
  const ev = { task_id, type };
  if (name !== undefined) ev.name = name;
  if (key !== undefined) ev.key = key;
  if (table !== undefined) ev.table = table;
  if (keys !== undefined) ev.keys = keys;
  if (value !== undefined) ev.value = value;
  if (fields && typeof fields === 'object') Object.assign(ev, fields);
  writeEvent(ev);
}

function emitEvent(task_id, name, fields) {
  emit({ task_id, type: 'event.emit', name, fields });
}

function dbWrite(task_id, table, fields) {
  emit({ task_id, type: 'db.write', table, fields });
}

function taskReturn(task_id, keys, fields) {
  emit({ task_id, type: 'task.return', keys, fields });
}

function logEmit(task_id, key, fields) {
  emit({ task_id, type: 'log.emit', key, fields });
}

function metricEmit(task_id, name, value, fields) {
  emit({ task_id, type: 'metric.emit', name, value, fields });
}

module.exports = {
  emit,
  emitEvent,
  dbWrite,
  taskReturn,
  logEmit,
  metricEmit,
};

File: stacks/go/base/code/task-trace.tpl.go
package tasktrace

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

type Event map[string]any

func tracePath() string {
	p := os.Getenv("TASK_TRACE_PATH")
	if p == "" {
		p = filepath.Join("artifacts", "task-trace.jsonl")
	}
	return p
}

func Emit(event Event) error {
	p := tracePath()
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return err
	}
	if _, ok := event["ts"]; !ok {
		event["ts"] = float64(time.Now().UnixNano()) / 1e9
	}
	b, err := json.Marshal(event)
	if err != nil {
		return err
	}
	f, err := os.OpenFile(p, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.Write(append(b, '\n'))
	return err
}

func EmitEvent(taskID string, name string, fields Event) error {
	ev := Event{"task_id": taskID, "type": "event.emit", "name": name}
	for k, v := range fields {
		ev[k] = v
	}
	return Emit(ev)
}

func DBWrite(taskID string, table string, fields Event) error {
	ev := Event{"task_id": taskID, "type": "db.write", "table": table}
	for k, v := range fields {
		ev[k] = v
	}
	return Emit(ev)
}

func TaskReturn(taskID string, keys []string, fields Event) error {
	ev := Event{"task_id": taskID, "type": "task.return", "keys": keys}
	for k, v := range fields {
		ev[k] = v
	}
	return Emit(ev)
}

func LogEmit(taskID string, key string, fields Event) error {
	ev := Event{"task_id": taskID, "type": "log.emit", "key": key}
	for k, v := range fields {
		ev[k] = v
	}
	return Emit(ev)
}

func MetricEmit(taskID string, name string, value any, fields Event) error {
	ev := Event{"task_id": taskID, "type": "metric.emit", "name": name, "value": value}
	for k, v := range fields {
		ev[k] = v
	}
	return Emit(ev)
}
